packages/fields/src/actions/set-as-homepage.tsx

```diff
@@ -0,0 +1,247 @@
+/**
+ * WordPress dependencies
+ */
+import { select, useDispatch } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import type { Settings, Page } from '@wordpress/core-data';
+import { __, sprintf } from '@wordpress/i18n';
+import {
+	Button,
+	__experimentalText as Text,
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+} from '@wordpress/components';
+import type { Action, ActionModal } from '@wordpress/dataviews';
+import { store as noticesStore } from '@wordpress/notices';
+
+/**
+ * Internal dependencies
+ */
+import { getItemTitle } from './utils';
+import type { CoreDataError, PostWithPermissions } from '../types';
+
+const PAGE_POST_TYPE = 'page';
+
+const SetAsHomepageModal: ActionModal< PostWithPermissions >[ 'RenderModal' ] =
+	( { items, closeModal, onActionPerformed } ) => {
+		const [ item ] = items;
+		const pageTitle = getItemTitle( item );
+		const siteSettings = select( coreStore ).getEntityRecord< Settings >(
+			'root',
+			'site'
+		);
+		const pageOnFront = siteSettings?.page_on_front || null;
+		const currentHomePage =
+			pageOnFront &&
+			select( coreStore ).getEntityRecord(
+				'postType',
+				'page',
+				pageOnFront
+			);
+		const pageForPosts = siteSettings?.page_for_posts;
+		const showOnFront = siteSettings?.show_on_front;
+		const currentHomePageTitle = getItemTitle( currentHomePage as Page );
+
+		const { editEntityRecord, saveEditedEntityRecord, saveEntityRecord } =
+			useDispatch( coreStore );
+		const { createSuccessNotice, createErrorNotice } =
+			useDispatch( noticesStore );
+
+		async function onSetPageAsHomepage( event: React.FormEvent ) {
+			event.preventDefault();
+
+			try {
+				// If selected page is set to draft, publish the page.
+				if ( item.status === 'draft' ) {
```

Comment:
I see that this was suggested, but I feel like a notice should be shown that the page will be published right ahead..

Reply:
Good idea. For now, I've made it more clear that a draft page is being published in the modal that's shown when the action is clicked:

![image](https://github.com/user-attachments/assets/be686658-7d5a-4573-9928-5274e53df0c7)